### ISSUE TITLE
Feature Request: Include an option to Assume Role in auth_login with AWS method

### ISSUE THREAD

**Issue by VictorCovalski** [hashicorp/terraform-provider-vault#1187]
# Context

In AWS, IAM Roles are a way to grant cross account access to authorized principals. This is achieved by making an API call to the Security Token Service (STS) Action [`AssumeRole`](). Upon success, the caller receives back temporary STS credentials  (Access Key ID, Secret Access Key and Session Token) which can be used to make further API calls to AWS with the identity and permissions of the assumed role.

The Vault [AWS IAM Auth](https://www.vaultproject.io/api/auth/aws) method works by binding Auth Roles to AWS Unique Ids. In order to obtain an IAM User/Role Unique ID from the ARN, Vault invokes `GetRole` or `GetUser` IAM actions. Therefore, an IAM Role must exist when an Auth Role is created in Vault. 

In order to overcome this limitation, Vault administrators may bind Auth Roles to IAM Roles that can be assumed from authorized principals throughout the organization. An example would be:

```shell
creds=`aws sts assume-role --role-arn arn:aws:iam::123456789:role/MyRole --role-session-name VaultSession | jq .Credentials`

export AWS_ACCESS_KEY_ID=`jq -r .AccessKey <<< $creds`
export AWS_SECRET_ACCESS_KEY=`jq -r .SecretKey <<< $creds`
export AWS_SESSION_TOKEN=`jq -r .SessionToken <<< $creds`

vault login -method=aws role=my-vault-role
```

# Proposal
Include an option that allows users of the Vault provider to specify a Role ARN that Terraform will assume before authenticating to Vault.

## Example:

```hcl
provider vault {
  address = "http://localhost:8200"
  auth_login {
    path = "auth/aws/login"
    method = "aws"
    parameters = {
      role = "vault"
      role_arn = "arn:aws:iam::123456789:role/MyRole"
    }
  }
}
```

# Draft
I have already developed a first working version of what I'm proposing here. Please take a look at https://github.com/VictorCovalski/terraform-provider-vault/commit/95213082fd794cb35ea6e0f6fe060f1eba30e840

Please share your opinions! Thanks!

**Comment by benashz** [comment]
In [v3.9.0](https://github.com/hashicorp/terraform-provider-vault/releases/tag/v3.9.0) we added first-class login support for the `aws` method which should resolve this issue. Please take a look at  [auth_login_aws](https://registry.terraform.io/providers/hashicorp/vault/latest/docs#auth_login_aws) for more details.

**Comment by chornbec** [comment]
I'm hoping to contribute some documentation on `auth_login_aws` because I'm trying to use the feature described in this feature request.
TL;DR, despite my provider statement below, the provider is still trying to use the credentials in my ~.aws/credentials file.
```
provider "vault" {
  address = "http://127.0.0.1:8200"
  skip_child_token = true
  auth_login_aws {
    role = "dev-role-corey-iam"
    header_value = "vault.example.com"
    aws_role_arn = "arn:aws:iam::1234567890:role/SomeRole"
    aws_role_session_name = "corey-test-session"
    aws_region = "us-east-1"
  }
}
```

I'm just trying to get Terraform to pull secrets from a local Vault instance using the `SomeRole` role, which is bound to the `dev-role-corey-iam` Auth role. I'm currently logged in with a different AWS IAM role (`PrimaryRole`) that can assume the `SomeRole` role. When I run `terraform plan`, the provider is using `PrimaryRole` instead when I expect it to run sts:AssumeRole so it would use `SomeRole`. Am I misinterpreting what `aws_role_arn` should do?

**Comment by Dragubitt** [comment]
Hello :) 

I understand the same about `aws_role_arn` parameter.. Did you finally manage to get it work ? 

Thanks

**Comment by chornbec** [comment]
No :( the docs aren't too helpful. For what it's worth, I'm using Leapp to manage my AWS credentials. Leapp basically modifies the `~/.aws/credentials` file to add/remove profiles and their keys, so I don't think using Leapp is causing `aws_role_arn` to not work.

**Comment by Dragubitt** [comment]
Oh ok nice, i did not know Leapp ;)
On my side I have no aws files, since i'm running my TF jobs on a EC2 instance. I would like to use instance profile and their authentication metadatas instead but it does not work : 

```shell
│ Error: Missing required argument
│ 
│   with provider["registry.terraform.io/hashicorp/vault"],
│   on terraform.tf line 3, in provider "vault":
│    3:   auth_login_aws {
│ 
│ "auth_login_aws.0.aws_secret_access_key": all of
│ `auth_login_aws.0.aws_access_key_id,auth_login_aws.0.aws_secret_access_key`
│ must be specified
```

Even if the doc is talking about "Optionnal" parameters : https://registry.terraform.io/providers/hashicorp/vault/latest/docs#aws

**Comment by chornbec** [comment]
>I would like to use instance profile and their authentication metadatas instead but it does not work

Not to deviate too much from this thread, but perhaps your EC2 instance needs `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` environment variables set so Terraform can use those. The `auth_login_aws` parameter looks for those env vars, which can be set after the EC2 instance retrieves temporary security credentials using its instance profile. [Guide](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html#instance-metadata-security-credentials). Perhaps if you create a script that sets those env vars after getting the credentials, your Terraform plan would work.

**Comment by whathelll** [comment]
I've raised https://github.com/hashicorp/terraform-provider-vault/issues/1754 regarding these access_key_id being optional but not optional
